test/e2e/autoscale_test.go

```diff
@@ -200,57 +202,66 @@ func assertScaleDown(ctx *testContext) {
 	ctx.t.Log("Scaled down.")
 }
 
-func TestAutoscaleUpDownUp(t *testing.T) {
-	t.Parallel()
-	ctx := setup(t)
-	defer test.TearDown(ctx.clients, ctx.names)
-
-	sc := startDiagnosis(t, ctx.clients, 15*time.Second)
-	defer sc.stop()
-
-	assertScaleUp(ctx)
-	assertScaleDown(ctx)
-	assertScaleUp(ctx)
-}
-
-func assertNumberOfPods(ctx *testContext, numReplicasMin int32, numReplicasMax int32) error {
-	deployment, err := ctx.clients.KubeClient.Kube.AppsV1().Deployments(test.ServingNamespace).Get(ctx.deploymentName, metav1.GetOptions{})
+func numberOfPods(ctx *testContext) (int32, error) {
+	deployment, err := ctx.clients.KubeClient.Kube.Apps().Deployments(test.ServingNamespace).Get(ctx.deploymentName, metav1.GetOptions{})
 	if err != nil {
-		return errors.Wrapf(err, "Failed to get deployment %q", deployment)
-	}
-	gotReplicas := deployment.Status.ReadyReplicas
-	mes := fmt.Sprintf("got %d replicas, expected between [%d, %d] replicas for deployment %s", gotReplicas, numReplicasMin, numReplicasMax, ctx.deploymentName)
-	ctx.t.Log(mes)
-	if gotReplicas < numReplicasMin || gotReplicas > numReplicasMax {
-		return errors.New(mes)
+		return 0, errors.Wrapf(err, "Failed to get deployment %q", deployment)
 	}
-	return nil
+	return deployment.Status.ReadyReplicas, nil
 }
 
-func assertAutoscaleUpToNumPods(ctx *testContext, numPods int32) {
-	// Relaxing the pod count requirement a little bit to avoid being too flaky.
-	minPods := numPods - 1
-	maxPods := numPods + 1
+func assertAutoscaleUpToNumPods(ctx *testContext, curPods, targetPods int32, duration time.Duration, quick bool) {
+	// There are two test modes: quick, and not quick.
+	// 1) Quick mode: succeeds when the number of pods meets targetPods.
+	// 2) Not Quick (sustaining) mode: succeeds when the number of pods gets scaled to targetPods and
+	//    sustains there for the `duration`.
+
+	// Relax the bounds to reduce the flakiness caused by sampling in the autoscaling algorithm.
+	minPods := curPods - 1
+	maxPods := targetPods + 1
 
 	stopChan := make(chan struct{})
 	var grp errgroup.Group
 	grp.Go(func() error {
-		return generateTraffic(ctx, int(numPods*10), 60*time.Second, stopChan)
+		return generateTraffic(ctx, int(targetPods*containerConcurrency), duration + 2*time.Second, stopChan)
```

Comment:
What is `2s` for?

Reply:
When the duration is over, we check the number of pods again.  The 2s extra is to make sure that the traffic is still sustaining for the last check.  It is an effort to reduce the flakiness of pods getting lower than the target, but doesn't seem effective.  Removing it to avoid confusion.